drivers: spi: silabs: eusart: Asynchronous support

flapla opened the pull request:
The goal of this pull request is to add Asynchronous transfer while in lower energy mode support to silabs SPI eusart driver.
This pull request does the following:

- Add DMA transfer to the silabs SPI eusart async function.
- Replace the default implementation of the SPI transfer function with the asynchronous implementation when CONFIG_SPI_ASYNC is set.
- Set up power management so that the system can lower energy mode during transfers.

Commit:
drivers: spi: silabs: eusart: Asynchronous support

Add support for asynchronous transfer in silabs eusart spi driver.

Signed-off-by: Francois Laplante <frlaplan@silabs.com>

Martinhoff-maker reviewed (changes requested):
Good !
Don't forget to add Silicon Laboratories Inc at the top of the file 👍 
Some comments can be applied on the whole code (for the declaration, blank spaces, initialization etc...).
Last question: have you tried to run the spi_loopback zephyr test that seems quite generic ? if yes, you may want to add a commit where you add an overlay file that show users how to configure the spi in dts.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +211,295 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+}
+
+#define LDMA_SIGNAL_TO_DMA_SLOT_CONFIG(slot) \
```
The PR is merged.

Martinhoff-maker commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,6 +709,12 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
 };
 
+#if defined(CONFIG_SPI_ASYNC)
+#define  SPI_DMA_DEVICE_INSTANCE .dma_device = (const struct device *)DEVICE_DT_GET_ONE(silabs_ldma),
```
Can be a goal like for asynchronous usart but not necessary if you add an overlay file in the spi_loopback test case where you add dmas field only for the desired eusart instance (it also shows users how to use asynchronous spi Silabs driver). You can check the overlay for asynchronous usart test of how it's done.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +212,314 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+	if (is_tx) {
+		/* Memory to peripheral. */
+		current_dma_config->channel_direction = 0x02U;
```
Should this be `enum dma_channel_direction` instead of a magic number?
Though: `MEMORY_TO_PERIPHERAL` is 1 and `PERIPHERAL_TO_MEMORY` is 2, so this seems backwards?

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +212,314 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		current_dma_config->channel_direction = 0x02U;
+	} else {
+		/* Peripheral to memory. */
+		current_dma_config->channel_direction = 0x01U;
```
Same as above.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +212,314 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	current_dma_config->user_data = (void *)dev;
+
+	if (!is_tx) {
+		/* Only enable the end transaction callback. Only SPI master mode is supported. */
```
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language
Should be `SPI controller mode`.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +212,314 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+#define LDMA_SIGNAL_TO_DMA_SLOT_CONFIG(slot)                                                       \
+	((slot & DMA_SLOT_SOURCESEL_MASK) << 13 | (slot & DMA_SLOT_SIGSEL_MASK))
+
+static inline uint8_t spi_silabs_get_eusart_source_signal(uint32_t eusart_num, uint32_t dma_channel,
```
Should this be sourced from DTS instead of being hard-coded here?
Binding headers like https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/dma/silabs/xg24-dma.h contain all the signals for use in DTS, avoiding the need to do this at runtime.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,21 +738,28 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
-		.clock_frequency = DT_INST_PROP_OR(n, clock_frequency, 1000000)			\
-	};											\
+#if defined(CONFIG_SPI_ASYNC)
+#define SPI_DMA_DEVICE_INSTANCE .dma_device = (const struct device *)DEVICE_DT_GET_ONE(silabs_ldma),
```
DMA device should be gotten from `dmas` property of the instance, not globally using a compatible

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -252,28 +658,61 @@ static int spi_silabs_eusart_transceive(const struct device *dev, const struct s
+		spi_context_release(&data->ctx, ret);
+		return ret;
+	}
+#if CONFIG_SPI_ASYNC
```
I don't fully understand the purpose of this guard. `spi_context_wait_for_completion` will do the right thing and either wait using a semaphore (which gets posted by `spi_context_complete`) or immediately return if `ctx->asynchronous` is true. And that variable gets set by the argument to `spi_context_lock`.

There seems to in general be unneeded complexity between the implementations of sync and async transfers here -- most other implementations simply implement a single transceive function that takes a `bool async`.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -15,33 +16,68 @@
 
+/* Global variable */
+#ifdef CONFIG_SPI_ASYNC
+static volatile uint8_t dummy_byte = SPI_TX_DUMMY_BYTE_VALUE;
```
Should use `SPI_MOSI_OVERRUN_DT_INST`, which gets seeded from the `overrun-character` property in DTS.

teburd commented:
SPI_ASYNC has nothing to do with DMA, DMA is usable with the blocking transceive API. SPI_ASYNC is an optional API for getting a signal at the end of a transcieve rather than blocking the thread while the transcieve is going on.

I'd also note that SPI_ASYNC is not actually asynchronous if you look. There's a semaphore, and if the bus is busy with other transcieves your thread will block. Is this really what you wanted? If you want asynchronous transcieves without blocking the calling thread then you need to use RTIO and and the submit API which is *guaranteed to not block* and *work from any call context* which is likely what you really wanted in an async API.

jhedberg commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -1,11 +1,16 @@
 
 #define DT_DRV_COMPAT silabs_eusart_spi
 
+#if defined(CONFIG_SPI_ASYNC)
+#define SPI_SILABS_SPI_EUSART_DMA (1U)
+#endif
```
It looks to me like you're using this new define as a kind of feature switch, even though it's not a Kconfig option. I'd suggest to make it a proper Kconfig option instead. It can be promptless and auto-enabled whenever `CONFIG_SPI_ASYNC` is enabled, to keep the same semantics as here, i.e. something like the following in `Kconfig.silabs_eusart`:
```
config SPI_SILABS_EUSART_DMA
	depends on SPI_SILABS_EUSART
	default y if SPI_ASYNC
```

asmellby replied:
This description doesn't entirely match my suggestion, DMA should be usable independently of SPI_ASYNC=y.
What I tried to suggest is as follows: It is possible to use the presence of a `dmas` property in DTS to decide whether a _specific_ instance should use DMA or not. When it comes to globally enabling DMA _support_ in the driver, this can be done either through a Kconfig option as @jhedberg suggests or by looking at some DTS property (such as checking if _any_ SPI node has a `dmas` property).

Looking at drivers in the tree, it seems like _most_ drivers do this with Kconfig, while some drivers compile it in unconditionally and only use DTS to decide whether a specific driver instance should use it or not. My conclusion was therefore to go the Kconfig route.

jhedberg commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -15,6 +20,12 @@
 #include <zephyr/drivers/clock_control/clock_control_silabs.h>
 #include <zephyr/drivers/pinctrl.h>
 #include <zephyr/logging/log.h>
+#ifdef SPI_SILABS_SPI_EUSART_DMA
+#include <zephyr/drivers/dma.h>
+#endif
```
Unless you have a good justification to do so, conditional includes are not a good idea, since they prevent the possibility of taking advantage of `IS_ENABLED()` in the code.

jhedberg commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -23,25 +34,56 @@ LOG_MODULE_REGISTER(spi_silabs_eusart, CONFIG_SPI_LOG_LEVEL);
 #define SPI_WORD_SIZE 8
 
-/* Structure Declarations */
+#if	defined(CONFIG_SPI_ASYNC) && defined(SPI_SILABS_SPI_EUSART_DMA)
```
One of these conditions seems redundant since you're always enabling SPI_SILABS_SPI_EUSART_DMA together with CONFIG_SPI_ASYNC. Once you've added the new Kconfig option this should just be `#if defined(CONFIG_SPI_SILABS_EUSART_DMA)`

jhedberg commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -23,25 +34,56 @@ LOG_MODULE_REGISTER(spi_silabs_eusart, CONFIG_SPI_LOG_LEVEL);
+/* Structure Declarations */
 struct spi_silabs_eusart_data {
 	struct spi_context ctx;
+#if defined(CONFIG_SPI_ASYNC) && defined(SPI_SILABS_SPI_EUSART_DMA)
```
Same here. I won't comment in all places, but please fix them

jhedberg commented on soc/silabs/silabs_s2/Kconfig.defconfig:
```diff
@@ -18,4 +18,10 @@ config SILABS_SLEEPTIMER_TIMER
 config CORTEX_M_SYSTICK
 	default n if SILABS_SLEEPTIMER_TIMER || GECKO_BURTC_TIMER
 
+config ZERO_LATENCY_IRQS
+	default y
+
+config ZERO_LATENCY_LEVELS
+	default 2
```
Please add some comments here explaining why these are needed.

jhedberg replied:
Yes, seems like that happened in #85519

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,21 +773,40 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
+	PM_DEVICE_DT_INST_DEFINE(n, spi_silabs_eusart_power_management_action);                \
+	SPI_DEVICE_DT_INST_DEFINE(n, spi_silabs_eusart_init, PM_DEVICE_DT_INST_GET(n),         \
+				  &spi_silabs_eusart_data_##n, &spi_silabs_eusart_cfg_##n,         		   \
+				  POST_KERNEL, CONFIG_SPI_INIT_PRIORITY, &spi_silabs_eusart_api);
```
Indent with tabs (1 tab == aligned on 8 spaces) but align with spaces.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	pm_device_runtime_put(dev);
+}
+
+#if defined(CONFIG_PM_DEVICE)
```
See https://github.com/zephyrproject-rtos/zephyr/pull/85386 -- the PM action should be unconditionally used together with `pm_device_driver_init()` for all flavors of PM to work correctly:
* CONFIG_PM
* CONFIG_PM_DEVICE + CONFIG_PM_DEVICE_SYSTEM_MANAGED
* CONFIG_PM_DEVICE + CONFIG_PM_DEVICE_RUNTIME

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -52,24 +91,26 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		 * No need to do it again, but must re-enable in case
+		 * TXEN/RXEN were cleared due to deep sleep.
+		 */
+#if !defined(CONFIG_PM_DEVICE_RUNTIME)
```
This doesn't seem right. In runtime device PM mode, the device lifetime is decoupled from the system lifetime, meaning that this mode should be _more_ likely to need to re-enable the EUSART after deep sleep compared to system managed device PM.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+}
+
+static inline size_t spi_silabs_longuest_transfer_size(struct spi_context *instance_ctx)
```
typo: longest

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+			dma_block_config_idx++;
+
+			if (dma_block_config_idx > SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
```
Can this condition occur? The while loop runs while `<`, and the variable is only incremented once. Seems like this check should be for `==`, and happen at the beginning of the loop, since `==` isn't an error condition when checked _after_ incrementing.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +217,330 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	while (padding_size) {
+
+		padding_size = spi_silabs_build_dma_transfer_block(
+			instance_conf, &current_blk_cfg[dma_block_config_idx], NULL, padding_size,
```
`dma_block_config_idx` could already be `== SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL` here, so this access is an overflow. The check on line 463 needs to happen at the beginning of the loop, before the variables are accessed.

asmellby replied:
I see this was modified, but now `dma_block_config_idx` is never incremented in the loop, so it will fail if more than 1 padding block is needed.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,21 +825,43 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
+	PM_DEVICE_DT_INST_DEFINE(n, spi_silabs_eusart_power_management_action); \
+	SPI_DEVICE_DT_INST_DEFINE(n, spi_silabs_eusart_init, PM_DEVICE_DT_INST_GET(n), \
+				             &spi_silabs_eusart_data_##n, &spi_silabs_eusart_cfg_##n, \
+				             POST_KERNEL, CONFIG_SPI_INIT_PRIORITY, &spi_silabs_eusart_api);
```
Align with opening parenthesis.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,21 +825,43 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
+		SPI_CONTEXT_INIT_SYNC(spi_silabs_eusart_data_##n, ctx), \
+		SPI_CONTEXT_CS_GPIOS_INITIALIZE(DT_DRV_INST(n), ctx) \
+		SPI_DMA_CHANNEL(n, rx) \
+		SPI_DMA_CHANNEL(n, tx) \
```
Prefix your macros with `SPI_SILABS_EUSART`. Otherwise, the reader could think `SPI_DMA_CHANNEL()` is provided by the SPI framework.

BTW, `SPI_SILABS_EUSART` is probably too long. I believe the name of the driver (`EUSART`) would be sufficient. However, this should be done in an atomic way, probably in a separate PR.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -298,21 +825,43 @@ static DEVICE_API(spi, spi_silabs_eusart_api) = {
+#else
+#define SPI_DMA_CHANNEL(index, dir)
+#define DMA_ENABLED_INSTANCE(n)
+#endif
```
`CONFIG_SPI_SILABS_EUSART_DMA` is no more required since you use `COND_CODE_1()`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
 }
 
-#ifdef CONFIG_SPI_ASYNC
+#if defined(CONFIG_SPI_ASYNC) 
```
I believe we tend to prefer `#ifdef` rather than `#if defined`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
+		/** Async without dma is not implmented, please configure the device tree
+		 *  instance with the proper dma configuration.
+		 */
+		return -ENOTSUP;
```
Context is not released.

BTW, locate all the clean-up code in a unique place:

```
	if (ret) {
		goto out;
	}
[...]
out:
	spi_context_release(&data->ctx, ret);
	return ret;
```

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
+
+	spi_silabs_eusart_xfer_dma(dev, config);
+
+	ret = spi_context_wait_for_completion(ctx);
```
I am not sure to understand what we are trying to do. The idea is to have an asynchronous function, right?

asmellby replied:
The function name is misleading, if the context is async, this helper function does _not_ wait. It effectively turns into a no-op, so it's always safe to call.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
+	struct spi_context *ctx = &data->ctx;
+
+	uint16_t control = 0;
+	int ret = 0U;
```
So, we are explicitly declaring an unsigned value to assign it to a signed variable. What is the point?

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
+					const struct spi_buf_set *rx_bufs)
 {
 	struct spi_silabs_eusart_data *data = dev->data;
+	const struct spi_silabs_eusart_config *eusart_config = dev->config;
```
Prefer to place the longest declaration first (reverse christmas tree rule).

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
 {
 	struct spi_silabs_eusart_data *data = dev->data;
+	const struct spi_silabs_eusart_config *eusart_config = dev->config;
 	uint16_t control = 0;
```
If the value of `control` is not meaningful, don't initialize it.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+
+	return ret;
+}
```
This function is definitely too long (100LoC) and too complex (5 level of indentation, 12 local variables). Normal humans can't juggle with more than 7 variables. Another symptom that this function is too complex: the length of the local variables.

jerome-pouiller replied:
(I have just noticed the other reviewers did the same comment)

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	current_dma_config->complete_callback_en = 0x00U;
+
+	current_dma_config->source_data_size = 0x01U; /* 1 byte. */
+	current_dma_config->dest_data_size = 0x01U;   /* 1 byte. */
```
Why do we need to write `0x01U` rather than `1`?

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	eusart->CMD_SET = EUSART_CMD_CLEARTX;
+
+	while (EUSART_StatusGet(eusart) & EUSART_STATUS_RXFL) {
+		(void)eusart->RXDATA;
```
Even if Cortex-M is not tricky, I suggest to use accessors to access to hardware registers:
```
sys_read32(&eusart->RXDATA);
```
These function, add the necessary memory barriers (on compiler and on CPU sides).

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		(void)eusart->RXDATA;
+	}
+
+	while (eusart->STATUS & EUSART_STATUS_CLEARTXBUSY) {
```
What is purpose of `EUSART_StatusGet()`?

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_SLEEP);
```
What is the purpose of assigning the result to `ret` if you don't test it?

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +227,343 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+}
+
+static int spi_silabs_eusart_power_management_action(const struct device *dev,
+						     enum pm_device_action action)
```
It would help if the the PM was in a separated commit.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,10 +181,41 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
 			       (clock_control_subsys_t)&eusart_config->clock_cfg);
-	if (err < 0) {
+
+	if (err < 0 && err != -EALREADY) {
```
This change is not related to this commit. Place it in a separate commit.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,10 +181,41 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		}
+
+		data->dma_channel_tx.channel_nb = dma_request_channel(data->dma_channel_tx.dma_device,
+																NULL);
```
Indentation

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -37,11 +64,26 @@ struct spi_silabs_eusart_config {
+	volatile uint8_t mosi_overrun_val;
+#if defined(CONFIG_SPI_SILABS_EUSART_DMA)
+	const bool is_dma_enabled_instance;
+#endif
```
Do not duplicate the information. This field is redundant with `dma_channel_rx`/`dma_channel_tx`.

jerome-pouiller commented:
(I have not yet finished my review. I will continue tomorrow.)

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,82 +617,194 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+					      void *userdata)
 {
-	return -ENOTSUP;
+	struct spi_silabs_eusart_data *data = dev->data;
```
There is still unnecessary duplication between the polling and async implementations. You should be able to implement this as:
```c
int transceive(dev, config, tx, rx, async, cb, userdata)
{
    /* The actual work goes here in a single implementation */
}

int transceive_polling(dev, config, tx, rx)
{
    return transceive(dev, config, tx, rx, false, NULL, NULL);
}

#ifdef CONFIG_SPI_ASYNC
int transceive_async(dev, config, tx, rx, cb, userdata)
{
    return transceive(dev, config, tx, rx, true, cb, userdata);
}
#endif
```

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,82 +617,194 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
 
+	err = pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0 && (err != -ENOENT)) {
+		/** The return number ENOENT means that no sleep state is defined
```
Yes, `pm_device_driver_init()` calls the `RESUME` action in its stub implementation when PM is disabled. This is why you need to use that function as the last step in init (`return pm_device_driver_init(dev);`), and why the pm_action callback needs to be unconditionally compiled into the driver.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -37,11 +63,31 @@ struct spi_silabs_eusart_config {
+
+	return data->dma_channel_rx.dma_dev != NULL;
+}
+#endif
```
I suggest to avoid conditional declaration of a function (note I am not talking about the definition). The caller will have to place the call under `#ifdef`. I suggest this alternative:
```
static inline bool is_dma_enabled_instance(const struct device *dev)
{
#ifdef CONFIG_SPI_SILABS_EUSART_DMA
	struct spi_silabs_eusart_data *data = dev->data;

	return data->dma_channel_rx.dma_dev != NULL;
#else
	return false;
#endif
}
```

This should allow to simplify some lines of code.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -52,24 +98,26 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		 * No need to do it again, but must re-enable in case
+		 * TXEN/RXEN were cleared due to deep sleep.
+		 */
+#if !defined(CONFIG_PM_DEVICE_RUNTIME)
```
I believe you can avoid the `#ifdef` here:
```
if (!IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
	EUSART_Enable(eusart_config->base, eusartEnable);
}
```

asmellby replied:
You added the `#if !defined(CONFIG_PM_DEVICE_RUNTIME)` guard around the existing code when moving it. That's what I'm referring to. The code (as it was) is still needed, consider the following usage pattern:

```c
struct device *dev = DT_ALIAS(my_spi);

pm_device_runtime_get(dev);
spi_transceive(dev, ...);
k_msleep(100); /* or anything else that suspends the thread and causes a deep sleep, such as pending on a semaphore */
spi_transceive(dev, ...);
pm_device_runtime_put(dev);
```

In this scenario, the first transfer will have configured the EUSART to the appropriate settings. After sleep, the second transfer will hit this early-exit condition when reconfiguring for the next transfer. However, since EUSART loses its enable state in EM2, it needs to be unconditionally re-enabled even if the configuration is kept and doesn't need repainting.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +232,344 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	pm_policy_state_lock_get(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+	pm_device_runtime_get(dev);
```
@ceolin @bjarki-andreasen @JordanYates Double checking this: Prior to https://github.com/zephyrproject-rtos/zephyr/issues/84049 going through, is my above statement ("drivers should not take their own PM handle") true? Or should this SPI driver take its own PM handle until the RFC passes and all subsystems that sit on top are updated to do it?

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +597,150 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+	struct spi_context *ctx = &data->ctx;
+	int ret;
+
+	if (asynchronous) {
```
You should be able to just pass the function arguments directly to `spi_context_lock`, you know this function is only called from two places and that `cb` and `userdata` are correctly `NULL` whenever `asynchronous` is `false`.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -236,45 +749,49 @@ static int spi_silabs_eusart_init(const struct device *dev)
-	struct spi_silabs_eusart_data *data = dev->data;
-	uint16_t control = 0;
-	int ret;
+	pm_device_init_off(dev);
```
Don't call this function, it's done inside `pm_device_driver_init()` if the system is configured for it.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -236,45 +749,49 @@ static int spi_silabs_eusart_init(const struct device *dev)
+	pm_device_init_off(dev);
 
-	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+	err = pm_device_runtime_enable(dev);
```
Don't call this function, it's done inside `pm_device_driver_init()` if the system is configured for it.

asmellby replied:
For context: when using `pm_device_driver_init`, device runtime PM is automatically enabled if the DTS node has the property `zephyr,pm-device-runtime-auto`. This allows the user to opt-in to runtime PM on a per-device basis (or in the future, opt-out when it becomes enabled by default). If we want to enable it by default for a given device, we can set the property in DTS and have users use `/delete-property/` to opt out. Setting `CONFIG_PM_DEVICE_RUNTIME=y` and _not_ setting `zephyr,pm-device-runtime-auto` means that the device is `RESUME`d at boot and never transitions again, effectively "not using device runtime PM". Since not all users of the SPI drivers currently deal with device runtime PM correctly (the spi_nor flash driver does it, but ls0xx display driver doesn't), IMO it's safest to leave runtime PM off in DTS, and allow users to opt in by setting it for their board given their specific usecase for a peripheral where they know they're `get`-ing and `put`-ing it correctly at the application level.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -236,45 +749,49 @@ static int spi_silabs_eusart_init(const struct device *dev)
-	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+	err = pm_device_runtime_enable(dev);
+	
+	pm_device_driver_init(dev, spi_silabs_eusart_power_management_action_pm_action);
```
The return value of `pm_device_driver_init()` must be returned from the init function.

asmellby replied:
Also typo: s/power_management_action_pm_action/pm_action/

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +230,326 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	pm_policy_state_lock_put(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+}
+
+static int spi_silabs_eusart_power_management_action_pm_action(const struct device *dev,
```
typo: s/power_management_action_pm_action/pm_action/

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +232,326 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+void spi_silabs_dma_rx_callback(const struct device *dev, void *user_data, uint32_t channel,
+				int status)
+{
+	if (status == DMA_STATUS_COMPLETE || status < 0U) {
```
I suggest to handle the the exception case first:
```
if (status >= 0 && status != DMA_STATUS_COMPLETE) {
    return;
}
```
Thus, the function is simpler and you save one level of indentation.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +232,326 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	pm_policy_state_lock_put(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+}
+
+static int spi_silabs_eusart_power_management_pm_action(const struct device *dev,
```
This is still an unnecessarily long name. Remove `power_management` from the name, `pm` is already there.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +599,144 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+		} else {
+			ret = spi_silabs_eusart_xfer_polling(dev, config);
+		}
+	}
```
Return value from `xfer` is not handled.

jhedberg commented:
@flapla note that there are three issues that the Compliance check is reporting

asmellby reviewed (commented):
Please add board overlays for the `spi_loopback` test for at least one Silabs board where DMA is configured. Please also add test cases for the ASYNC=n and DMA=n scenarios to exercise all the driver configurations.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +592,145 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+		}
+	}
+
+	ret = spi_context_wait_for_completion(ctx);
```
The DMA case already calls this within `spi_silabs_eusart_xfer_dma`, doing it again here doesn't seem right.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +592,145 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+
+	if (spi_context_wait_for_completion(&data->ctx) >= 0) {
+		/* Successful transaction */
+		return 0;
```
Agree. In the async case, early return is correct, since cleanup is done in the RX complete ISR. In the sync case, early return is correct when transfer succeeded, since cleanup is done in the RX complete ISR. If the transfer failed, it falls through to the cleanup here.

It still seems like the return value from `spi_context_wait_for_completion` should be captured and returned in that case. I.e. this should be something like
```c
ret = spi_context_wait_for_completion(&data->ctx);
if (ret >= 0) {
	/* Successful transaction */
	return 0;
}
/* continue to cleanup */
return ret;
```

asmellby commented on drivers/spi/Kconfig.silabs_eusart:
```diff
@@ -12,3 +12,13 @@ config SPI_SILABS_EUSART
+	select DMA
+	default y if SPI_ASYNC
+
+config SPI_DMA_MAX_TRANSFER_BLOCK_PER_CHANNEL
```
Should this option have `SILABS_EUSART` in it? Could be something like `SPI_SILABS_EUSART_DMA_MAX_BLOCKS`.
It should also depend on the above config such that it's only available if DMA is enabled.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +591,144 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+	ret = dma_start(data->dma_channel_rx.dma_dev, data->dma_channel_rx.channel_nb);
+
+	if (ret) {
+		dma_stop(data->dma_channel_tx.dma_dev, data->dma_channel_tx.channel_nb);
```
Did you mean to put this here? I would think it should go in the next `if (ret)` below, and stop RX, not TX.

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +646,147 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
 	struct spi_silabs_eusart_data *data = dev->data;
+	int err;
+
+	err = clock_control_on(eusart_config->clock_dev,
```
Does the clock need to be turned on in the init function? In the UART driver, this is needed because initial peripheral configuration happens in init. But since the SPI driver does peripheral configuration in the transaction function, it seems like clock enable could be deferred until the PM resume callback.

Commit:
test: drivers: spi: silabs: Add loopback test support.

Add silabs spi eusart driver Async/dma,no-dma test cases.
With required configuration files.

Signed-off-by: Francois Laplante <frlaplan@silabs.com>

Commit:
drivers: spi: silabs: eusart: Add pm support

Add device policy and runtime power management support in
silabs spi eusart driver.

Signed-off-by: Francois Laplante <frlaplan@silabs.com>

jerome-pouiller reviewed (commented):
Thank you for the split in smaller chunks. It helps a lot.

I have only reviewed the first commit for now. I hope to have time to review the other commits during the day.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -19,29 +19,64 @@
+
 LOG_MODULE_REGISTER(spi_silabs_eusart, CONFIG_SPI_LOG_LEVEL);
 
 #include "spi_context.h"
```
Indeed. A large part of `spi_context.h` should go in `spi_context.c`. Meanwhile, you can add a comment above `LOG_MODULE_REGISTER()` (eg. "Required by spi_context.h").

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -236,6 +241,61 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+		pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+
+		break;
```
`EUSART_Enable(eusart_config->base, eusartEnable)` during the resume. Is it expected?

jerome-pouiller replied:
So, the call to `clock_control_on()` is also useless?

jerome-pouiller replied:
Maybe you could add this comment in the code.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -37,11 +65,33 @@ struct spi_silabs_eusart_config {
 /* Helper Functions */
-static int spi_silabs_eusart_configure(const struct device *dev, const struct spi_config *config,
-			     uint16_t *control)
+static inline bool is_dma_enabled_instance(const struct device *dev)
```
Sure, this is not in the scope of this PR.

jhedberg commented:
@fpistm @jerome-pouiller @Martinhoff-maker could you try to mark all resolved conversations as resolved? (the "Resolve conversation" button). This PR is starting to be quite slow to load/render for me, and there's quite a lot of scrolling required to get to the bottom, so being diligent with the resolving may help at least a little.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +240,307 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+	current_dma_config->head_block = (struct dma_block_config *)channel->dma_descriptors;
+
+	current_dma_config->user_data = (void *)dev;
```
Converting a pointer to a `void *` does not require an explicit cast (while you keep the same const or volatile qualifications).

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -152,6 +240,307 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+					      &(instance_data->dma_chan_rx);
+	struct dma_block_iterator blk_it = { .blk_array = &(channel->dma_descriptors[0]),
+					     .block_idx = 0U,
+					     .current_total_transfer_size = 0U };
```
I believe you get me wrong. These functions returns the last used descriptor or NULL if the number of descriptors is exhausted. The list of descriptor has to be initialized by the caller:

```
static void spi_eusart_reset_desc(struct dma_channel *channel)
{
	int i;

	memset(channel->dma_descriptors, 0, sizeof(channel->dma_descriptors));
	for (i = 0; i < ARRAY_SIZE(channel->dma_descriptors) - 1; i++) {
		channel->dma_descriptors[i]->next_block = &channel->dma_descriptors[i + 1];
	}
}

static int spi_eusart_prepare_dma_transaction(const struct device *dev)
{
	const struct spi_silabs_eusart_config *cfg = dev->config;
	struct spi_silabs_eusart_data *data = dev->data;
	int total_tx_len = spi_context_total_tx_len(data->ctx);
	int total_rx_len = spi_context_total_rx_len(data->ctx);
	struct dma_block_config *desc;

	spi_eusart_reset_desc(data->dma_chan_tx);
	desc = spi_eusart_fill_data_desc(cfg, data->dma_chan_tx->dma_descriptors,
					 data->ctx->current_tx, data->ctx->tx_count, true);
	if (!desc) {
		return -ENOMEM;
	}
	desc = spi_eusart_fill_padding_desc(cfg, desc->next_block,
					    MAX(total_rx_len - total_tx_len, 0), true);
	if (!desc) {
		return -ENOMEM;
	}
	desc->next_block = NULL;

	spi_eusart_reset_desc(data->dma_chan_rx);
	desc = spi_eusart_fill_data_desc(cfg, data->dma_chan_rx->dma_descriptors,
					 data->ctx->current_rx, data->ctx->rx_count, false);
	if (!desc) {
		return -ENOMEM;
	}
	desc = spi_eusart_fill_padding_desc(cfg, desc->next_block,
					    MAX(total_tx_len - total_rx_len, 0), false);
	if (!desc) {
		return -ENOMEM;
	}
	desc->next_block = NULL;
	return 0;
}
```

Interesting fact: with this method, the descriptors could be easily shared between tx and rx channels.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +588,147 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+		return ret;
+	}
+
+	spi_silabs_eusart_pm_policy_get(dev);
```
I meant this function does not yet exist in this commit.

FWIW, I suggest to check each of your commit compiles with something like: `git rebase -i origin/main --exec "ninja -C build"`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,35 +588,147 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+
+force_transaction_close:
+	spi_context_cs_control(ctx, false);
+	spi_silabs_eusart_pm_policy_put(dev);
```
But this function did not exist in the first commit. This line has to be moved in the 3rd commit to keep the atomicity of the commits.

jerome-pouiller commented on drivers/spi/Kconfig.silabs_eusart:
```diff
@@ -12,3 +12,14 @@ config SPI_SILABS_EUSART
+	bool "Silabs EUSART SPI controller driver DMA enabled transfer"
+	depends on SPI_SILABS_EUSART
+	select DMA
+	default y if SPI_ASYNC
```
So, why not `default y` rather than `default y if SPI_ASYNC`?

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,25 +503,127 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+
+	ret = spi_silabs_eusart_configure(dev, config);
+
+	if (ret) {
```
Remove blank line

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -199,25 +503,127 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
+		} else {
+			ret = spi_silabs_eusart_xfer_polling(dev, config);
+		}
+	}
```
You can take advantage of `else if` and save one level of indentation:

```
	if (spi_silabs_eusart_is_dma_enabled_instance(dev)) {
		ret = spi_silabs_eusart_xfer_dma(dev, config);
	} else if (!asynchronous) {
		ret = spi_silabs_eusart_xfer_polling(dev, config);
	} else {
		/* Asynchronous transfers without DMA is not implemented,
		 * please configure the device tree
		 * instance with the proper DMA configuration.
		 */
		ret = -ENOTSUP;
	}
```

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	if (ret == -ENOMEM) {
+		LOG_ERR("DMA driver failed allocation.");
+		return ret;
+	}
```
I am not sure of the usefulness of this `LOG_ERR()`. Why for this function and not for the `spi_eusart_fill_data_desc()`?

jerome-pouiller replied:
You can drop the condition now.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		return -ENOMEM;
+	}
+
+	ret = spi_silabs_dma_config(spi_dev, channel, is_tx);
```
`ARRAY_INDEX(channel->dma_descriptors, desc)` will provide `block_count`. You can get rid of `spi_silabs_descriptor_chain_size()`

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		return 0U;
+	}
+
+	ret = spi_eusart_prepare_dma_transaction(dev, padded_transaction_size, true);
```
The two functions has the same names: `spi_silabs_prepare_dma_transaction()` vs `spi_eusart_prepare_dma_transaction()`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -286,33 +682,53 @@ static int spi_silabs_eusart_release(const struct device *dev, const struct spi_
+		.clock_cfg = SILABS_DT_INST_CLOCK_CFG(n), \
+		.mosi_overrun = (uint8_t)SPI_MOSI_OVERRUN_DT(n), \
+		.clock_frequency = DT_INST_PROP_OR(n, clock_frequency, 1000000), \
+		}; \
```
Indentation

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -241,40 +647,30 @@ static int spi_silabs_eusart_init(const struct device *dev)
+static int spi_silabs_eusart_transceive_sync(const struct device *dev,
+				     const struct spi_config *config,
+				     const struct spi_buf_set *tx_bufs,
+				     const struct spi_buf_set *rx_bufs)
```
Indentation

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+	if (padded_transaction_size == 0) {
+		/* Nothing to do */
+		return 0U;
```
But this function return an int.

TBH, I don't understand why some developers add `U` everywhere:
```
    int a = 0U;    /* No warning while this is missleading */
    char b = 128U;  /* No warnings unless -Wconversion. Behave the same than `b = 128` */
    bool c = (-1 > 0U); /* This expression return true (while "-1 > 0" obviously returns false) */

```

What is the justification?

jerome-pouiller reviewed (commented):
About the commit title, the prefix should be `drivers: spi: silabs-eusart:`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,288 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		}
+		offset += spi_eusart_fill_desc(cfg, desc,
+					       buffer[i].buf ? (uint8_t *)buffer[i].buf + offset
+								: NULL,
```
Ok, then I suggest place this expression in a variable.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -37,23 +58,39 @@ struct spi_silabs_eusart_config {
+static volatile uint8_t empty_buffer;
+#endif
+
+static inline bool spi_silabs_eusart_is_dma_enabled_instance(const struct device *dev)
```
Remove the `inline`.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,286 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+}
+
+#ifdef CONFIG_SPI_SILABS_EUSART_DMA
+void spi_silabs_dma_rx_callback(const struct device *dev, void *user_data, uint32_t channel,
```
`static`

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,286 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	spi_context_complete(instance_ctx, spi_dev, status);
+}
+
+static inline void spi_silabs_eusart_clear_txrx_fifos(EUSART_TypeDef *eusart)
```
Remove the `inline`

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,286 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+}
+
+static inline size_t spi_silabs_longest_transfer_size(struct spi_context *instance_ctx)
```
Remove the `inline`

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +174,286 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	return MAX(tx_transfer_size, rx_transfer_size);
+}
+
+static inline size_t spi_silabs_descriptor_chain_size(struct dma_block_config *desc)
```
I believe this function is orphan now.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -228,6 +231,58 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+		EUSART_Enable(eusart_config->base, eusartDisable);
+		ret = clock_control_off(eusart_config->clock_dev,
+					(clock_control_subsys_t)&eusart_config->clock_cfg);
+
```
Remove this line

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -228,6 +231,58 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+		if (ret == -EALREADY) {
+			/* Ignore if clock is already on */
+			ret = 0U;
```
`ret` is signed.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -228,6 +231,58 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+		if (ret == -EALREADY) {
+			/* Ignore if clock is already on */
+			ret = 0U;
```
`ret` is signed.

jerome-pouiller commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +177,339 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+					(clock_control_subsys_t)&eusart_config->clock_cfg);
+
+		if (ret == -EALREADY) {
+			/* Ignore if clock is already on */
```
`Ignore if clock is already off` (BTW, I suggest to drop this comment rather than confusing the reader).

asmellby commented on drivers/spi/spi_silabs_eusart.c:
```diff
@@ -137,21 +177,323 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+
+		if (data->dma_chan_tx.chan_nb < 0) {
+			dma_release_channel(data->dma_chan_rx.dma_dev, data->dma_chan_rx.chan_nb);
+			data->dma_chan_rx.chan_nb = 0;
```
Should be `-1` since it's unallocated?
